src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.cs

```diff
@@ -463,93 +423,170 @@ protected void InitializeReferences()
             }
         }
 
+        /// <summary>
+        /// Performs the matching against a given string input, which will call the passed in <paramref name="callback"/> for every
+        /// match that is found, along with the relevant <paramref name="state"/>.
+        /// </summary>
         internal void Run<TState>(string input, int startat, ref TState state, MatchCallback<TState> callback, bool reuseMatchObject)
         {
             Debug.Assert((uint)startat <= (uint)input.Length);
 
             RegexRunner runner = Interlocked.Exchange(ref _runner, null) ?? CreateRunner();
             try
             {
-                runner.InitializeTimeout(internalMatchTimeout);
+                // For the string overload, we need to set runtext before starting the match attempts.
                 runner.runtext = input;
-                int runtextpos = startat;
-                while (true)
-                {
-                    runner.InitializeForScan(this, input, startat, false);
-                    runner.runtextpos = runtextpos;
+                RunInternal(input, startat, ref state, callback, runner, usingStringOverload: true, reuseMatchObject);
+            }
+            finally
+            {
+                runner.runtext = null; // drop reference to text to avoid keeping it alive in a cache.
+                _runner = runner;
+            }
+        }
+
+        /// <summary>
+        /// Performs the matching against a given span input, which will call the passed in <paramref name="callback"/> for every
+        /// match that is found, along with the relevant <paramref name="state"/>.
+        /// </summary>
+        internal void Run<TState>(ReadOnlySpan<char> input, int startat, ref TState state, MatchCallback<TState> callback, bool reuseMatchObject)
+        {
+            Debug.Assert((uint)startat <= (uint)input.Length);
+
+            RegexRunner runner = Interlocked.Exchange(ref _runner, null) ?? CreateRunner();
+            try
+            {
+                RunInternal(input, startat, ref state, callback, runner, usingStringOverload: false, reuseMatchObject);
+            }
+            finally
+            {
+                _runner = runner;
+            }
```

Comment:
This change is adding even more layers of function calls to get from entry into the guts of the matching, and the presence of the try/finally means this method won't be inlined (though even without the try/finally it's likely it's still too big to inline by default).  We should double-check that this isn't negatively impacting throughput for the methods that go through here.